ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,390 @@
+    long segmentCapacity = firstSegment.getCapacityInBytes();
+    // adding the left over capacity in lower segment + capacities of all segments inbetween + offset in higher segment
+    long difference =
+        segmentCapacity - lower.getOffset() + segmentCapacity * interveningSegmentsCount + higher.getOffset();
+    return o1.compareTo(o2) * difference;
   }
 
-  @Override
-  public void appendFrom(ReadableByteChannel channel, long size)
-      throws IOException {
-    logger.trace("Log : {} currentWriteOffset {} capacityInBytes {} sizeToAppend {}", file.getAbsolutePath(),
-        currentWriteOffset, capacityInBytes, size);
-    if (currentWriteOffset.get() + size > capacityInBytes) {
-      metrics.overflowWriteError.inc(1);
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to append to log " +
-          "from channel since new data size " + size + "exceeds total log size " + capacityInBytes);
-    }
-    long bytesWritten = 0;
-    while (bytesWritten < size) {
-      bytesWritten += fileChannel.transferFrom(channel, currentWriteOffset.get() + bytesWritten, size - bytesWritten);
-    }
-    currentWriteOffset.addAndGet(bytesWritten);
-    logger.trace("Log : {} bytes appended to the log from read channel bytesWritten: {}", file.getAbsolutePath(),
-        bytesWritten);
-  }
-
-  /**
-   * Writes {@code size} number of bytes from the channel {@code channel} into the log at {@code offset}.
-   * @param channel The channel from which data needs to be written from.
-   * @param offset The offset in the segment at which to start writing.
-   * @param size The amount of data in bytes to be written from the channel.
-   * @throws IOException if data could not be written because of I/O errors
-   *
+  /**
+   * Flushes the Log and all its segments.
+   * @throws IOException if the flush encountered an I/O error.
    */
-  void writeFrom(ReadableByteChannel channel, long offset, long size)
-      throws IOException {
-    logger.trace("Log : {} currentWriteOffset {} capacityInBytes {} sizeToAppend {} offset to append at {}",
-        file.getAbsolutePath(), currentWriteOffset, capacityInBytes, size, offset);
-    if (offset < 0 || offset + size > currentWriteOffset.get()) {
-      metrics.overflowWriteError.inc(1);
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to write to log " +
-          "from channel since new data size " + size + "exceeds log end offset " + currentWriteOffset.get());
+  void flush() throws IOException {
+    // TODO: try to flush only segments that require flushing.
+    logger.trace("Flushing log");
+    for (LogSegment segment : segmentsByName.values()) {
+      segment.flush();
     }
-    long bytesWritten = 0;
-    while (bytesWritten < size) {
-      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+  }
+
+  /**
+   * Closes the Log and all its segments.
```

Comment:
I assumed that filechannel's close will provide a flush also but that is probably a bad assumption. Will flush explicitly.